src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/DefaultModelBindingContext.cs

```diff
@@ -0,0 +1,324 @@
+            }
+        }
+
+        /// <inheritdoc />
+        public override Type ModelType => ModelMetadata?.ModelType;
+
+        /// <inheritdoc />
+        public override string BinderModelName
+        {
+            get { return _state.BinderModelName; }
+            set { _state.BinderModelName = value; }
+        }
+
+        /// <inheritdoc />
+        public override BindingSource BindingSource
+        {
+            get { return _state.BindingSource; }
+            set { _state.BindingSource = value; }
+        }
+
+        /// <inheritdoc />
+        public override Type BinderType
+        {
+            get { return _state.BinderType; }
+            set { _state.BinderType = value; }
+        }
+
+        /// <inheritdoc />
+        public override bool FallbackToEmptyPrefix
+        {
+            get { return _state.FallbackToEmptyPrefix; }
+            set { _state.FallbackToEmptyPrefix = value; }
+        }
+
+        /// <inheritdoc />
+        public override bool IsTopLevelObject
+        {
+            get { return _state.IsTopLevelObject; }
+            set { _state.IsTopLevelObject = value; }
+        }
+
+        /// <inheritdoc />
+        public override IValueProvider ValueProvider
+        {
+            get { return _state.ValueProvider; }
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+                _state.ValueProvider = value;
+            }
+        }
+
+        /// <inheritdoc />
+        public override Func<ModelBindingContext, string, bool> PropertyFilter
+        {
+            get { return _state.PropertyFilter; }
+            set { _state.PropertyFilter = value; }
+        }
+
+        /// <inheritdoc />
+        public override ValidationStateDictionary ValidationState
+        {
+            get { return _state.ValidationState; }
+            set { _state.ValidationState = value; }
+        }
+
+        /// <inheritdoc />
+        public override ModelBindingResult? Result
+        {
+            get
+            {
+                return _state.Result;
+            }
+            set
+            {
+                if (value.HasValue && value.Value == default(ModelBindingResult))
+                {
+                    throw new ArgumentException(nameof(ModelBindingResult));
+                }
+
+                _state.Result = value;
+            }
+        }
+
+
+        private struct State
+        {
+            public OperationBindingContext OperationBindingContext;
+            public string FieldName;
+            public object Model;
+            public ModelMetadata ModelMetadata;
+            public string ModelName;
+
+            public IValueProvider ValueProvider;
+            public Func<ModelBindingContext, string, bool> PropertyFilter;
+            public ValidationStateDictionary ValidationState;
+            public ModelStateDictionary ModelState;
```

Comment:
OBC, ModelState, and ValidationState are really global things, you can't scope em. Removing these from `State` would trim a few bytes.